src/skottie-player.ts

```diff
@@ -141,54 +440,368 @@ export class SkottiePlayer extends LitElement {
+            reader.onload = () => {
+              this.setContent(JSON.parse(reader.result as string));
+              this.assets = {};
+              this.imageColors = {};
+              this.fileString = JSON.stringify(this.content);
+              this._loadJSONEditor();
+              this.updatePlayer();
+              this.fileUploaded = true;
+            };
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Handles asset upload. Updates assets, color, and player based on
+   * file content.
+   */
+  handleAssetUpload(e: Event): void {
+    const assetInput = (e.composedPath()[0] as HTMLInputElement);
+    const mapKey = assetInput.getAttribute('map-key');
+    const assetType = assetInput.getAttribute('asset-type')
+    if (assetInput && mapKey && assetType) {
+      const files = assetInput.files;
       if (files) {
         const file = files[0];
-        const reader = new FileReader();
+        if (file) {
+          const reader = new FileReader();
+          
+          if (assetType === 'image') {
+            Promise.all([
+              createImageBitmap(file),
+            ]).then((imageBitmaps) => {
+              (this.imageColors as any)[mapKey] = colorThief.getColor(imageBitmaps[0]);
+            });
+          }
 
-        this.title = file.name;
-        this.fileUrl = window.URL.createObjectURL(file);
+          reader.readAsArrayBuffer(file);
+          reader.onload = () => {
+            (this.assets as any)[mapKey] = reader.result;
+            this.updatePlayer();
+          }
+        }
+      }
+    }
+  }
 
-        reader.readAsText(file);
-        reader.onload = () => {
-          this.setContent(JSON.parse(reader.result as string));
-          lottiePlayer.destroy();
-          this.updatePlayer();
-        };
+  /**
+   * Loads content into the JSONEditor.
+   */
+  _loadJSONEditor(): void {
+    (this.jsonEditor as JSONEditor).set(this.content);
+  }
+
+  /**
+   * Shows the JSONEditor and hids the GUI editor. Also applies updates
+   * to the stored JSON data.
+   */
+  showJSONEditor(): void {
+    this.updateJson();
+    this.showJsonEditor = true;
+    this._loadJSONEditor();
+  }
+
+  /**
+   * Shows the GUI editor and hides the JSON Editor. Also applies updates
+   * to the stored JSON data.
+   */
+  hideJSONEditor(): void {
+    this.updateJson();
+    this.showJsonEditor = false;
+  }
+
+  /**
+   * Updates the downloadable json file, content, and animation based on 
+   * changes in the editor.
+   */
+  updateJson(): void {
+      // pull updated json
+      if (this.showJsonEditor) {
+        this._updateJsonWithEditor();
+      } else {
+        this._updateJsonFramerate();
+        this._updateJsonLayers();
+      }
+      this.fileString = JSON.stringify(this.content);
+
+      // update download link
+      URL.revokeObjectURL(this.fileUrl);
+      this.fileUrl = URL.createObjectURL(new Blob([this.fileString]));
+      
+      // update animation
+      lottiePlayer.destroy();
```

Comment:
Is lottiePlayer still used?

Also, we should probably deallocate the prev animation at this point: animation.delete()

Reply:
lottiePlayer isn't used anymore. I updated the code to delete the old animation whenever a new animation is created.